desktop_server/src/com/mygdx/game/server/util/TechnologyUtilServer.java

```diff
@@ -101,30 +107,43 @@ private void applyTechnologyToExistingEntities(int technologyEntityId, Owner own
             applyTechnologyToUnit(technologyConfig, entityId);
         }
       }
-      case MATERIAL_IMPACT -> {
-        // TODO
-      }
       case BUILDING_IMPACT -> {
+        for (int i = 0; i < underConstructionSubscriber.getEntities().size(); i++) {
+          int entityId = underConstructionSubscriber.getEntities().get(i);
+          int parent = subfieldMapper.get(underConstructionMapper.get(entityId).getParentSubfield()).getParent();
+          if (ownerMapper.get(parent).getToken().equals(owner.getToken()))
+            applyTechnologyToBuilding(technologyConfig, entityId);
+        }
+      }
+      case MATERIAL_IMPACT -> {
         // TODO
       }
     }
   }
 
-  public void applyTechnologyToNewEntities(int unitEntityId, TechnologyImpactType technologyType) {
-    var owner = ownerMapper.get(unitEntityId);
+  public void applyTechnologyToNewEntities(int newEntityId, String ownerToken, TechnologyImpactType technologyType) {
     var entities = researchedOwnerSubscriber.getEntities();
     switch (technologyType) {
       case UNIT_IMPACT -> {
         for (int i = 0; i < entities.size(); i++) {
           int entityId = entities.get(i);
-          if (gameConfigAssets.getGameConfigs().get(TechnologyConfig.class, entityConfigIdMapper.get(entityId).getId()).getImpact().getTechnologyImpactType().equals(technologyType) && ownerMapper.get(entityId).getToken().equals(owner.getToken())) {
-            log.info("Apply technology " + entityConfigIdMapper.get(entityId).getId() + " to unit " + unitEntityId);
-            applyTechnologyToUnit(gameConfigAssets.getGameConfigs().get(TechnologyConfig.class, entityConfigIdMapper.get(entityId).getId()), unitEntityId);
+          if (gameConfigAssets.getGameConfigs().get(TechnologyConfig.class, entityConfigIdMapper.get(entityId).getId()).getImpact().getTechnologyImpactType().equals(technologyType)
```

Comment:
😰😰😰
teraz to już nie zmieniaj jak może robisz coś innego, ale takie coś mogłoby być rozdzielone do zmiennych

Reply:
Poprawię w następnym z technologii 🖌️